src/libostree/ostree-repo-composefs.c

```diff
@@ -0,0 +1,631 @@
+  const guint64 source_size = g_file_info_get_size (source_info);
+  const char *source_symlink_target = g_file_info_get_symlink_target (source_info);
+  const gboolean is_symlink
+      = (g_file_info_get_file_type (source_info) == G_FILE_TYPE_SYMBOLIC_LINK);
+
+  struct lcfs_node_s *node = lcfs_node_new ();
+  if (node == NULL)
+    return glnx_throw (error, "Out of memory");
+
+  /* Takes ownership on success */
+  if (lcfs_node_add_child (parent, node, destination_name) != 0)
+    {
+      lcfs_node_unref (node);
+      return glnx_throw_errno (error);
+    }
+
+  lcfs_node_set_mode (node, source_mode);
+  lcfs_node_set_uid (node, source_uid);
+  lcfs_node_set_gid (node, source_gid);
+  lcfs_node_set_size (node, source_size);
+  if (is_symlink)
+    {
+      if (lcfs_node_set_payload (node, source_symlink_target) != 0)
+        return glnx_throw_errno (error);
+    }
+  else if (source_size != 0)
+    {
+      char loose_path_buf[_OSTREE_LOOSE_PATH_MAX];
+      _ostree_loose_path (loose_path_buf, checksum, OSTREE_OBJECT_TYPE_FILE, OSTREE_REPO_MODE_BARE);
+      if (lcfs_node_set_payload (node, loose_path_buf) != 0)
+        return glnx_throw_errno (error);
+
+      guchar *known_digest = NULL;
+
+#ifdef HAVE_LINUX_FSVERITY_H
+      /* First try to get the digest directly from the bare repo file.
+       * This is the typical case when we're pulled into the target
+       * system repo with verity on and are recreating the composefs
+       * image during deploy. */
+      char buf[sizeof (struct fsverity_digest) + OSTREE_SHA256_DIGEST_LEN];
+
+      if (G_IS_UNIX_INPUT_STREAM (input))
+        {
+          int content_fd = g_unix_input_stream_get_fd (G_UNIX_INPUT_STREAM (input));
+          struct fsverity_digest *d = (struct fsverity_digest *)&buf;
+          d->digest_size = OSTREE_SHA256_DIGEST_LEN;
+
+          if (ioctl (content_fd, FS_IOC_MEASURE_VERITY, d) == 0
+              && d->digest_size == OSTREE_SHA256_DIGEST_LEN
+              && d->digest_algorithm == FS_VERITY_HASH_ALG_SHA256)
+            known_digest = d->digest;
+        }
+#endif
+
+      if (known_digest)
+        lcfs_node_set_fsverity_digest (node, known_digest);
+      else if (lcfs_node_set_fsverity_from_content (node, input, _composefs_read_cb) != 0)
+        return glnx_throw_errno (error);
+    }
+
+  if (xattrs)
+    {
+      if (!_ostree_composefs_set_xattrs (node, xattrs, cancellable, error))
+        return FALSE;
+    }
+
+  g_clear_object (&input);
+
+  return TRUE;
+}
+
+static gboolean
+checkout_composefs_recurse (OstreeRepo *self, const char *dirtree_checksum,
+                            const char *dirmeta_checksum, struct lcfs_node_s *parent,
+                            const char *name, GCancellable *cancellable, GError **error)
+{
+  g_autoptr (GVariant) dirtree = NULL;
+  g_autoptr (GVariant) dirmeta = NULL;
+  g_autoptr (GVariant) xattrs = NULL;
+  struct lcfs_node_s *directory;
```

Comment:
Here and everywhere - in newer ostree code we prefer C99-style declare-and-initialize.